src/dbnode/encoding/tile/series_frame_iterator_test.go

```diff
@@ -0,0 +1,200 @@
+			frames[i] = v
+			v++
+		}
+	}
+
+	return frames
+}
+
+func halfFrameCounts(numPoints int) []int {
+	frames := make([]int, numPoints*2-1)
+	for i := range frames {
+		if i%2 == 0 {
+			frames[i] = 1
+		}
+	}
+
+	return frames
+}
+
+func TestSeriesFrameIterator(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	numPoints := 30
+	start := time.Now().Truncate(time.Hour)
+
+	stepSize := time.Second * 10
+
+	tests := []struct {
+		name      string
+		frameSize time.Duration
+		exCounts  []int
+		exSums    []float64
+	}{
+		{
+			name:      "5 second frame, 1 point every 2 frames - with empty frames",
+			frameSize: time.Second * 5,
+			exSums:    halfFrameSizes(numPoints),
+			exCounts:  halfFrameCounts(numPoints),
+		},
+		{
+			name:      "1 minute frame, 6 points per frame",
+			frameSize: time.Minute * 1,
+			exSums: []float64{
+				15 /* Σ 0..5 */, 51 /* Σ 6..11 */, 87, /* Σ 12..17 */
+				123 /*Σ18..23 */, 159 /*Σ24..30 */},
+			exCounts: []int{6, 6, 6, 6, 6},
+		},
+		{
+			frameSize: time.Minute * 2,
+			exSums:    []float64{66 /* Σ 0..11 */, 210 /* Σ 12..23 */, 159 /*Σ24..30 */},
+			exCounts:  []int{12, 12, 6},
+		},
+		{
+			name:      "3 minute frame, 18 points per frame",
+			frameSize: time.Minute * 3,
+			exSums:    []float64{153 /*Σ0..17 */, 282 /* Σ 18..30 */},
+			exCounts:  []int{18, 12},
+		},
+		{
+			name:      "4 minute frame, 24 points per frame",
+			frameSize: time.Minute * 4,
+			exSums:    []float64{276 /*Σ0..23 */, 159 /* Σ 24..30 */},
+			exCounts:  []int{24, 6},
+		},
+		{
+			name:      "5 minute frame, 30 points per frame",
+			frameSize: time.Minute * 5,
+			exSums:    []float64{435},
+			exCounts:  []int{30},
+		},
+		{
+			name:      "6 minute frame, 30 points per frame (exhausted)",
+			frameSize: time.Minute * 6,
+			exSums:    []float64{435},
+			exCounts:  []int{30},
+		},
+	}
+
+	recorder := newFlatDatapointRecorder()
+	it := newSeriesFrameIterator(recorder)
+	require.False(t, it.Next())
+	require.Error(t, it.Err())
+
+	for _, tt := range tests {
+		iter := newSequentialIterator(ctrl, start, stepSize, numPoints)
+		require.NoError(t, it.Reset(
+			xtime.ToUnixNano(start),
+			xtime.UnixNano(tt.frameSize),
+			iter,
+		))
+
+		step := 0
+		exVal := 0.0
+		exTime := start.UnixNano()
+		for it.Next() {
+			require.True(t, step < len(tt.exSums))
+			frame := it.Current()
+			assert.NotNil(t, frame)
+			assert.Equal(t, tt.exSums[step], frame.Sum())
+
+			vals := frame.Values()
+			require.Equal(t, tt.exCounts[step], len(vals))
+			for i := 0; i < tt.exCounts[step]; i++ {
+				assert.Equal(t, exVal, vals[i])
+				exVal++
+			}
+
+			times := frame.Timestamps()
+			require.Equal(t, tt.exCounts[step], len(times))
+			for i := 0; i < tt.exCounts[step]; i++ {
+				assert.Equal(t, exTime, times[i].UnixNano())
+				exTime = exTime + int64(time.Second*10)
+			}
+
+			step++
+		}
+
+		assert.Equal(t, len(tt.exSums), step)
+		assert.NoError(t, it.Err())
+	}
+	assert.NoError(t, it.Close())
```

Comment:
Should be inside test table loop.

Reply:
This one closes the parent iterator rather than the internal iterators, but did need a close within the loop too